Fetch token via certificate (#206)
promregator/promregator: docs/config.md

java -Dspring.config.location=file:/path/to/your/myconfig.yaml -jar promregator-0.0.1-SNAPSHOT.jar
```

> #### Option "promregator.authenticator.oauth2xsuaa.tokenServiceURL" (mandatory, if using promregator.authenticator.type=OAuth2XSUAABasic)

Author: I "simply" did not document the deprecated way. Of course the deprecated way still works, but I omitted it here in order to keep the docu simple. _Maybe I did not get your point (?)_

> #### Option "promregator.authenticator.oauth2xsuaa.tokenServiceURL" (mandatory, if using promregator.authenticator.type=OAuth2XSUAABasic)

Reviewer: I think we are talking at cross purposes here. I'll try to give you a call tomorrow afternoon. Perhaps it's easier there to clarify that aspect here (I believe my intention is mainly trivial).

> #### Option "promregator.authenticator.oauth2xsuaa.tokenServiceURL" (mandatory, if using promregator.authenticator.type=OAuth2XSUAABasic)

Author: > I'll try to give you a call tomorrow afternoon.

👍

Specifies the URL of the OAuth2 endpoint, which contains the token service of your authorization server in case of global authentication. Typically, this is the endpoint with the path `/oauth/token`, as Promregator will try to perform to establish a ["Client Credentials"-based authentication](https://www.digitalocean.com/community/tutorials/an-introduction-to-oauth-2#grant-type-client-credentials).

#### Option "promregator.authenticator.oauth2xsuaa.tokenServiceCertURL (mandatory for if using promregator.authenticator.type=OAuth2XSUAACertificate)"
Specifies the URL of the OAuth2 endpoint for certificate based authentication, which contains the token service of your authorization server in case of global authentication.

#### Option "promregator.authenticator.oauth2xsuaa.client_id" (mandatory, if using promregator.authenticator.type=OAuth2XSUAABasic or OAuth2XSUAACertificate)
Specifies the client identifier (a.k.a. "client_id") which shall be used during the OAuth2 request based on the Grant Type Client Credentials flow in case of global authentication.

#### Option "client_certificates" (mandatory if using promregator.authenticator.type=OAuth2XSUAACertificate)

> Specifies the certificate chain which shall be used during the OAuth2 request. Omitted when client secret-based authentification is used.

Author: The `omitted` part has been removed for `client_certificates´, `client_secret` and `client_key` since these parts are already implied by `mandatory if ...`.

```

#### Option "promregator.authenticator.oauth2xsuaa.tokenServiceURL" (mandatory, if using promregator.authenticator.type=OAuth2XSUAA, *deprecated*)
Specifies the URL of the OAuth2 endpoint, which contains the token service of your authorization server in case of global authentication. Typically, this is the endpoint with the path `/oauth/token`, as Promregator will try to perform to establish a ["Client Credentials"-based authentication](https://www.digitalocean.com/community/tutorials/an-introduction-to-oauth-2#grant-type-client-credentials). Deprecated, please use `promregator.authenticator.oauth2xsuaaBasic.tokenServiceURL` instead.

#### Option "promregator.authenticator.oauth2xsuaaBasic.tokenServiceURL" (mandatory, if using promregator.authenticator.type=OAuth2XSUAABasic)
Specifies the URL of the OAuth2 endpoint, which contains the token service of your authorization server in case of global authentication. Typically, this is the endpoint with the path `/oauth/token`, as Promregator will try to perform to establish a ["Client Credentials"-based authentication](https://www.digitalocean.com/community/tutorials/an-introduction-to-oauth-2#grant-type-client-credentials).

> #### Option "promregator.authenticator.oauth2xsuaaCertificate.tokenServiceCertURL" (mandatory for if using promregator.authenticator.type=OAuth2XSUAACertificate)

Author: Would be possible to change the naming here to `promregator.authenticator.oauth2xsuaaCertificate.tokenServiceURL` (... no `Cert`).

Then it would be the same like the other  `tokenServiceURL`s. The `Cert` is implied by the prefix.

As always in live there are pros an cons (... this the current approach it is better visible that it is something else).

In case I should change it I can adjust. @eaglerainbow : just let me know ...

Reviewer: I get your argument, but given the fact that this typically is also a different server and hence a different URL, I prefer `tokenServiceCertURL` (with Cert).
Perhaps a brief comment along the lines of 

> Note that this attribute is called `tokenServiceCertURL` and not `tokenServiceURL`!

might be helpful, though.

Author: done.

Specifies the URL of the OAuth2 endpoint for certificate-based authentication, which contains the token service of your authorization server in case of global authentication. Typically, this is the endpoint with the path `/oauth/token`, as Promregator will try to perform to establish a ["Client Credentials"-based authentication](https://www.digitalocean.com/community/tutorials/an-introduction-to-oauth-2#grant-type-client-credentials).

> #### Option "promregator.authenticator.oauth2xsuaa.client_id" (mandatory, if using promregator.authenticator.type=OAuth2XSUAA, *deprecated*)

Reviewer: Argl. Ordering here is mixed up! Note that we have first all `promregator.authenticator.oauth2xsuaa`, then `"promregator.authenticator.oauth2xsuaaBasic` would come next, followed by `"promregator.authenticator.oauth2xsuaaCertificate`.

Author: Done, documentation has been reordered accordingly.

*WARNING!*
Due to security reasons, it is *neither* recommended to store this value in your YAML file, nor to put it into the command line when starting Promregator.
Instead it is suggested to set the corresponding environment variables `PROMREGATOR_AUTHENTICATOR_OAUTH2XSUAA_CLIENT_SECRET` when starting the application.

Example:

```bash
export PROMREGATOR_AUTHENTICATOR_OAUTH2XSUAA_CLIENT_SECRET=myClientSecret
java -Dspring.config.location=file:/path/to/your/myconfig.yaml -jar promregator-0.0.1-SNAPSHOT.jar
```

#### Option "promregator.authenticator.oauth2xsuaaBasic.client_secret" (mandatory, if using promregator.authenticator.type=OAuth2XSUAABasic)
Specifies the client secret (a.k.a. "client_secret") which shall be used during the OAuth2 request based on the Grant Type Client Credentials flow in case of global authentication.

*WARNING!*
Due to security reasons, it is *neither* recommended to store this value in your YAML file, nor to put it into the command line when starting Promregator.
Instead it is suggested to set the corresponding environment variables `PROMREGATOR_AUTHENTICATOR_OAUTH2XSUAA_BASIC_CLIENT_SECRET` when starting the application.

Example:

```bash

> export PROMREGATOR_AUTHENTICATOR_OAUTH2XSUAA_BASIC_CLIENT_SECRET=myClientSecret

Reviewer: Is `PROMREGATOR_AUTHENTICATOR_OAUTH2XSUAA_BASIC_CLIENT_SECRET` really correct? I would have expected `PROMREGATOR_AUTHENTICATOR_OAUTH2XSUAABASIC_CLIENT_SECRET`.

Author: It works with these environment variables which are also contained in the docu like this:

- `PROMREGATOR_AUTHENTICATOR_OAUTH2XSUAA_BASIC_CLIENT_SECRET`
- `PROMREGATOR_AUTHENTICATOR_OAUTH2XSUAA_CLIENT_SECRET`
- `PROMREGATOR_AUTHENTICATOR_OAUTH2XSUAA_CERTIFICATE_CLIENT_KEY`

It works with the underscore between `OAUTH2XSUAA` and `BASIC`.

java -Dspring.config.location=file:/path/to/your/myconfig.yaml -jar promregator-0.0.1-SNAPSHOT.jar
```

#### Option "promregator.authenticator.oauth2xsuaaCertificate.client_key" (mandatory, if using promregator.authenticator.type=OAuth2XSUAACertificate)
Specifies the private key which shall be used during the OAuth2 request.

*WARNING!* 
Due to security reasons, it is *neither* recommended to store this value in your YAML file, nor to put it into the command line when starting Promregator.

> Instead it is suggested to set the corresponding environment variables `PROMREGATOR_AUTHENTICATOR_OAUTH2XSUAA_CLIENT_KEY` when starting the application.

Reviewer: ```suggestion
Instead it is suggested to set the corresponding environment variables `PROMREGATOR_AUTHENTICATOR_OAUTH2XSUAACERTIFICATE_CLIENT_KEY` when starting the application.
```

Example:

```bash

> export PROMREGATOR_AUTHENTICATOR_OAUTH2XSUAA_CERTFICATE_CLIENT_KEY=myClientKey

Reviewer: see question above about the additional underscore.

Author: answered above.

#### Option "promregator.authenticator.oauth2xsuaa.tokenServiceURL" (mandatory, if using promregator.authenticator.type=OAuth2XSUAA)
Specifies the URL of the OAuth2 endpoint, which contains the token service of your authorization server in case of global authentication. Typically, this is the endpoint with the path `/oauth/token`, as Promregator will try to perform to establish a ["Client Credentials"-based authentication](https://www.digitalocean.com/community/tutorials/an-introduction-to-oauth-2#grant-type-client-credentials).

> #### Option "promregator.authenticator.oauth2xsuaa.url"

Author: I introduced `url` here instead continuing with `tokenServiceURL` since the naming in side the service key files issued by cf is `url` for client-secret based cases and `cert-url` for certificate based cases.

Of course it is also possible to continue with `tokenServiceURL` and `tokenServiceCertificateURL`.

The old `tokenServiceURL` is still fully supported for the client-secret based case.

Reviewer: Argl. Things are getting now tricky:
* On the one hand "url" could be anything - for me this way too generic (you could also fill `https://github.com` in here, because that's also "some URL").
* On the other hand, there is "standard set by SAP CF" through there service key files which exactly uses this "weak" attribute name (presumably due to historical reasons).

What we need to ensure from our side at either way is that `tokenServiceURL` still works in a "compatibility" mode - we can set this attribute to "deprecated" and suggest not to use it anymore, but use `url` instead (a good example how we document deprecation can be found at `promregator.endpoint.maxProcessingTime`). We can also add a corresponding comment in the release note later to make people more aware of this change.

BTW: Using `tokenServiceURL` and then having the value only of the XSUAA server's main URL would also be wrong. So, if we wanted to properly type the URL, then the better suffix would be "xsuaaServiceURL" or so (I am open to alternative suggestions). 

Thinking aloud, here is one more idea about this: We could so this:
1. If `xsuaaServiceURL` is set, fine, take it (whatever state `tokenServiceURL` is then).
2. If `xsuaaServiceURL` is not set, but `tokenServiceURL` is, we try to strip `/oauth/token` from its end, which brings us to the value I would expect `xsuaaServiceURL` would have. If stripping fails, we raise an error (preventing startup) saying that there is a configuration mismatch, which we couldn't resolve and kindly ask the user to fix. At the same message, we can suggest to make use of the new `xsuaaServiceURL` parameter, dropping `tokenServiceURL`. 

We had similar cases already before, where he had written a "Promregator Note" in form of a Wiki page in this repo here. We then mentioned the URL of that wiki page then in the error message, guiding the user to the documentation (s)he now needs to read for details.

wdyt?

Author: I adjusted
- url -> xsuaaServiceURL
- cert_url -> xsuaaServiceCertURL

Reason why I started with that was: these terms (`url`, `certurl`) are used inside the service keys and also [here](https://github.com/SAP/cloud-security-xsuaa-integration/blob/fbe88ed48afd21b8f13717619770f9e06abc6cdf/java-api/src/main/java/com/sap/cloud/security/config/cf/CFConstants.java#L17) and [here](https://github.com/SAP/cloud-security-xsuaa-integration/blob/fbe88ed48afd21b8f13717619770f9e06abc6cdf/java-api/src/main/java/com/sap/cloud/security/config/cf/CFConstants.java#L40). But I don't have problems with adhering to the project conventions, of course.

The behaviour is now: when there is a `xsuaaServiceURL` already set a (deprecated) `tokenServiceURL` is not taken into account. Problem here: we depend on the order how the properties are not. Also not that cool ...
In case there is a `tokenServiceURL` which will not be ignored we strip a trailing `oauth/token`.
When we fail with getting a token we catch that exception and log it alongside with a hint pointing to the parameters which have been configured.

@eaglerainbow : Please let me know if that makes sense from your point of view. Otherwise I can adjusted again.

Like `promregator.authenticator.oauth2xsuaa.tokenServiceURL`, but without trailing `/oauth/token`.

#### Option "promregator.authenticator.oauth2xsuaa.cert_url"
Specifies the URL of the OAuth2 endpoint for certificate based authentication, which contains the token service of your authorization server in case of global authentication. Needs to be provided without a path.

#### Option "promregator.authenticator.oauth2xsuaa.client_id" (mandatory, if using promregator.authenticator.type=OAuth2XSUAA)
Specifies the client identifier (a.k.a. "client_id") which shall be used during the OAuth2 request based on the Grant Type Client Credentials flow in case of global authentication.

> #### Option "promregator.authenticator.oauth2xsuaa.client_secret" (mandatory for client secret based authentication, if using promregator.authenticator.type=OAuth2XSUAA)
> Specifies the client secret (a.k.a. "client_secret") which shall be used during the OAuth2 request based on the Grant Type Client Credentials flow in case of global authentication. Omitted when certificate based authentication is used.
> 

Reviewer: ```suggestion
#### Option "promregator.authenticator.oauth2xsuaa.client_secret" (mandatory for client_secret-based authentication, if using promregator.authenticator.type=OAuth2XSUAA)
Specifies the client secret (a.k.a. "client_secret") which shall be used during the OAuth2 request based on the Grant Type Client Credentials flow in case of global authentication. Omitted, if certificate based authentication is used.

```

> #### Option "client_certificates" (mandatory for certificate based authentication, if using promregator.authenticator.type=OAuth2XSUA)

Author: Not sure about the naming. I used the plural here `client_certificates` since it is a certificate chain. Of course only the last element in the chain is really important. Would also be possible to use the singular `client_certificate` here.

Reviewer: We also have other cases here in the configuration which uses the plural form, e.g. `promregator.scraping.threads`  and `.scopes` although that is a single value (and not a YAML array). I'd say that we use the most common and best understandable term to express "what's behind that attribute". I assume that it the vast majority of cases, we have more than one certificate in there (due to the intermediate certificate). So, if that's true, I think it should be called `client_certificates` (plural).

BTW: We should briefly mention the separator approach between different values. I assume it's "concatenated string, no separation necessary, as the certificates have a suffix and prefix".
Are users allowed to use newlines (escaping them?) or is this prohibited?

Author: > Are users allowed to use newlines (escaping them?) or is this prohibited?

The users should simply use what is provided in the service key files. There we have a long string containing all the certificates with the normal `-----BEGIN CERTIFICATE-----\n` and `-----END CERTIFICATE-----\n` The next `-----BEGIN CERTIFICATE-----\n` is right after `-----END CERTIFICATE-----\n`. I looks like: `...L5MSJYw==\n-----END CERTIFICATE-----\n-----BEGIN CERTIFICATE-----\nMIIGY...`

Of course we have also `\n` inbetween as usual.

> #### Option "client_certificates" (mandatory for certificate based authentication, if using promregator.authenticator.type=OAuth2XSUA)
> Specifies the certificate chain which shall be used during the OAuth2 request. Omitted when clientsecret based authentification is used.

Reviewer: ```suggestion
#### Option "client_certificates" (mandatory for certificate-based authentication, if using promregator.authenticator.type=OAuth2XSUAA)
Specifies the certificate chain which shall be used during the OAuth2 request. Omitted when client secret-based authentification is used.
```

> #### Option "client_key" (mandatory for certificate based authentication, if using promregator.authenticator.type=OAuth2XSUA)
> Specifies the private key which shall be used during the OAuth2 request. Omitted when clientsecret based authentification is used.

Reviewer: ```suggestion
#### Option "client_key" (mandatory for certificate based authentication, if using promregator.authenticator.type=OAuth2XSUAA)
Specifies the private key which shall be used during the OAuth2 request. Omitted when client secret-based authentification is used.
```

Reviewer: BTW: How about encryption of the PEM key here? Enforced unencrypted? May be encrypted using `spring-cloud` (I assume so)?

*WARNING!* 

> Due to security reasons, it is *neither* recommended to store this value in your YAML file, nor to put it into the command line when starting Promregator.

Reviewer: BTW: This and the next lines look somehow displaced now. They talk about `client_secret` stuff, but we are now in the section of `client_key`.

Author: fixed. Parameters `client_secet` and `client_key` are now right before the warning. In the warning the corresponding environment variables are not stated.

Author: The warning is now repeated with each parameter. I guess I can resolve this conversation.

#### Option "promregator.authenticator.oauth2xsuaa.tokenServiceURL" (mandatory, if using promregator.authenticator.type=OAuth2XSUAA)
Specifies the URL of the OAuth2 endpoint, which contains the token service of your authorization server in case of global authentication. Typically, this is the endpoint with the path `/oauth/token`, as Promregator will try to perform to establish a ["Client Credentials"-based authentication](https://www.digitalocean.com/community/tutorials/an-introduction-to-oauth-2#grant-type-client-credentials).

#### Option "promregator.authenticator.oauth2xsuaa.xsuaaServiceURL"

> Like `promregator.authenticator.oauth2xsuaa.tokenServiceURL`, but without trailing `/oauth/token`.

Reviewer: ```suggestion
Like `promregator.authenticator.oauth2xsuaa.tokenServiceURL`, but without trailing `/oauth/token`.

If using `promregator.authenticator.type=OAuth2XSUAA` either `promregator.authenticator.oauth2xsuaa.tokenServiceURL` or `promregator.authenticator.oauth2xsuaa.xsuaaServiceURL` must be provided. If both are provided, `promregator.authenticator.oauth2xsuaa.xsuaaServiceURL` takes precedence.
```

Reviewer: Sounds a little as if `tokenServiceURL` was still the "right option". I have a little the impression that `xsuaaServiceURL` is our favored approach in the meantime (see also replacements in examples below). So documentation tends into the "wrong" direction.

Author: The parameters `tokenServiceURL` and `xsuaaServiceURL` are very close to each other. So I would like to propose to deprecate `tokenServiceURL`. That means we enforce URLs like `https://example.org:1234` which are transformed internally to `https://example.org:1234/oauth/token`.

Other approach: It would also be possible to keep the `tokenServiceUrl` and to implement a `NullOAuth2ServiceEndpointsProvider` which does not modify the provided url(s). When we follow this approach we would have `tokenServiceUrl` and `tokenServiceCertUrl`. Should I implement it this way? Would be easy to implement.

> #### Option "promregator.authenticator.oauth2xsuaa.xsuaaServiceCertURL"

Reviewer: ```suggestion
#### Option "promregator.authenticator.oauth2xsuaa.xsuaaServiceCertURL (mandatory for certificate-based authentication, if using promregator.authenticator.type=OAuth2XSUAA)"
```

#### Option "promregator.authenticator.oauth2xsuaa.client_id" (mandatory, if using promregator.authenticator.type=OAuth2XSUAA)
Specifies the client identifier (a.k.a. "client_id") which shall be used during the OAuth2 request based on the Grant Type Client Credentials flow in case of global authentication.

#### Option "client_certificates" (mandatory for certificate-based authentication, if using promregator.authenticator.type=OAuth2XSUAA)
Specifies the certificate chain which shall be used during the OAuth2 request. Omitted when client secret-based authentification is used.

#### Option "promregator.authenticator.oauth2xsuaa.client_secret" (mandatory for client_secret-based authentication, if using promregator.authenticator.type=OAuth2XSUAA)
Specifies the client secret (a.k.a. "client_secret") which shall be used during the OAuth2 request based on the Grant Type Client Credentials flow in case of global authentication. Omitted, if certificate based authentication is used.

> #### Option "client_key" (mandatory for certificate based authentication, if using promregator.authenticator.type=OAuth2XSUAA)

Reviewer: ```suggestion
#### Option "promregator.authenticator.oauth2xsuaa.client_key" (mandatory for certificate based authentication, if using promregator.authenticator.type=OAuth2XSUAA)
```

Specifies the private key which shall be used during the OAuth2 request. Omitted when client secret-based authentification is used.

*WARNING!* 
Due to security reasons, it is *neither* recommended to store the value for `client_secret` and `client_key` in your YAML file, nor to put it into the command line when starting Promregator.

> Instead it is suggested to set the identically named environment variables `PROMREGATOR_AUTHENTICATOR_OAUTH2XSUAA_CLIENT_SECRET` and `PROMREGATOR_AUTHENTICATOR_OAUTH2XSUAA_CLIENT_KEY` when starting the application.

Reviewer: I think we have to split and duplicate this sentence here and also add it to the `client_secret` section.
There will be users out there which will only read one variant only.

Author: done
